core/src/main/scala/kafka/server/BrokerToControllerChannelManager.scala

```diff
@@ -0,0 +1,211 @@
+        false,
+        channelBuilder,
+        logContext
+      )
+      new NetworkClient(
+        selector,
+        manualMetadataUpdater,
+        config.brokerId.toString,
+        1,
+        0,
+        0,
+        Selectable.USE_DEFAULT_BUFFER_SIZE,
+        Selectable.USE_DEFAULT_BUFFER_SIZE,
+        config.requestTimeoutMs,
+        config.connectionSetupTimeoutMs,
+        config.connectionSetupTimeoutMaxMs,
+        ClientDnsLookup.DEFAULT,
+        time,
+        false,
+        new ApiVersions,
+        logContext
+      )
+    }
+    val threadName = threadNamePrefix match {
+      case None => s"broker-${config.brokerId}-to-controller-send-thread"
+      case Some(name) => s"$name:broker-${config.brokerId}-to-controller-send-thread"
+    }
+
+    new BrokerToControllerRequestThread(networkClient, manualMetadataUpdater, requestQueue, metadataCache, config,
+      brokerToControllerListenerName, time, threadName)
+  }
+
+  private[server] def sendRequest(request: RequestChannel.Request,
+                                  callback: RequestCompletionHandler): Unit = {
+    val requestBuilder = new NoOpRequestBuilder(request.context.header.apiKey, request.body[AbstractRequest])
+    requestQueue.put(BrokerToControllerQueueItem(requestBuilder, callback, ""))
+  }
+
+  private[server] def forwardRequest(originalRequest: RequestChannel.Request,
+                                     callback: RequestCompletionHandler): Unit = {
+    val requestBuilder = new NoOpRequestBuilder(originalRequest.context.header.apiKey, originalRequest.body[AbstractRequest])
+    requestQueue.put(BrokerToControllerQueueItem(requestBuilder, callback, originalRequest.context.principal.getName))
+  }
+}
+
+case class BrokerToControllerQueueItem(request: AbstractRequest.Builder[_ <: AbstractRequest],
+                                       callback: RequestCompletionHandler,
+                                       initialPrincipalName: String)
+
+class BrokerToControllerRequestThread(networkClient: KafkaClient,
+                                      metadataUpdater: ManualMetadataUpdater,
+                                      requestQueue: BlockingQueue[BrokerToControllerQueueItem],
+                                      metadataCache: kafka.server.MetadataCache,
+                                      config: KafkaConfig,
+                                      listenerName: ListenerName,
+                                      time: Time,
+                                      threadName: String)
+  extends InterBrokerSendThread(threadName, networkClient, time, isInterruptible = false) {
+
+  private var activeController: Option[Node] = None
+
+  override def requestTimeoutMs: Int = config.controllerSocketTimeoutMs
+
+  override def generateRequests(): Iterable[RequestAndCompletionHandler] = {
+    val requestsToSend = new mutable.Queue[RequestAndCompletionHandler]
+    if (requestQueue.peek() != null) {
+      val topRequest = requestQueue.take()
```

Comment:
Since `take` is blocking, do we need the `peek` call above? Actually, does InterBrokerSendThread expect `generateRequests` to block like this? Since it returns an iterable, I would guess that it doesn't. Maybe we should do something like:

```scala
val topRequest = requestQueue.poll()
if (topRequest != null) {
   // ...
}
```

Reply:
@mumrah : I think you're right that we should not block here.

@abbccdda: I don't think a blocking queue is the right abstraction here since we don't want to block.  How about just a lock and a list?

Reply:
I think changing to `poll` should be suffice?